src/lib/tls/tls_signature_scheme.cpp

```diff
@@ -242,10 +242,40 @@ AlgorithmIdentifier Signature_Scheme::key_algorithm_identifier() const noexcept
 
 AlgorithmIdentifier Signature_Scheme::algorithm_identifier() const noexcept
    {
-   auto emsa = EMSA::create(padding_string());
-   if(!emsa)
-      { return AlgorithmIdentifier(); }
-   return emsa->config_for_x509(algorithm_name(), hash_function_name());
+   switch(m_code)
+      {
+      case RSA_PKCS1_SHA1:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA3(SHA-1)"), AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA256:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA3(SHA-256)"), AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA384:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA3(SHA-384)"), AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA512:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA3(SHA-512)"), AlgorithmIdentifier::USE_NULL_PARAM);
+
+      case ECDSA_SHA1:
+         return AlgorithmIdentifier(OID::from_string("ECDSA/EMSA1(SHA-1)"), AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA256:
+         return AlgorithmIdentifier(OID::from_string("ECDSA/EMSA1(SHA-256)"), AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA384:
+         return AlgorithmIdentifier(OID::from_string("ECDSA/EMSA1(SHA-384)"), AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA512:
+         return AlgorithmIdentifier(OID::from_string("ECDSA/EMSA1(SHA-512)"), AlgorithmIdentifier::USE_EMPTY_PARAM);
+
+      case RSA_PSS_SHA256:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA4"),
+                                    hex_decode("3039A00F300D06096086480165030402010500A11C301A06092A864886F70D010108300D06096086480165030402010500A203020120A303020101"));
+      case RSA_PSS_SHA384:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA4"),
+                                    hex_decode("3039A00F300D06096086480165030402020500A11C301A06092A864886F70D010108300D06096086480165030402020500A203020130A303020101"));
+      case RSA_PSS_SHA512:
+         return AlgorithmIdentifier(OID::from_string("RSA/EMSA4"),
+                                    hex_decode("3039A00F300D06096086480165030402030500A11C301A06092A864886F70D010108300D06096086480165030402030500A203020140A303020101"));
+
+      default:
+         // Note that Ed25519 and Ed448 end up here
+         return AlgorithmIdentifier();
+      }
```

Comment:
Same as [last year](https://github.com/randombit/botan/pull/3125#discussion_r1049941524): Sad that we need to scatter those mappings around the code. But then again: it feels like every protocol has their own particularities that make it hard to abstract this.

Anyway: Maybe at least add a comment to the `hex_decode`-ed EMSA4 parameters explaining what's inside the magic strings?
In the linked comment, I regenerated those for every invocation. Less efficient but easier to understand:

```C++
   auto pss_params = [](const std::string hash_name)
      {
         const AlgorithmIdentifier hash_id(hash_name, AlgorithmIdentifier::USE_NULL_PARAM);
         const AlgorithmIdentifier mgf_id("MGF1", hash_id.BER_encode());

         std::vector<uint8_t> parameters;
         DER_Encoder(parameters)
            .start_sequence()
            .start_context_specific(0).encode(hash_id).end_cons()
            .start_context_specific(1).encode(mgf_id).end_cons()
            .start_context_specific(2).encode(HashFunction::create(hash_name)->output_length()).end_cons()
            .start_context_specific(3).encode(size_t(1)).end_cons() // trailer field
            .end_cons();

         return parameters;
      };
```

Still wondering whether `DER_Encoder` could do this as a `constexpr` generation. Purely academic question though.